src/kvstore/raftex/SnapshotManager.cpp

```diff
@@ -0,0 +1,116 @@
+/* Copyright (c) 2019 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+#include "kvstore/raftex/SnapshotManager.h"
+#include "base/NebulaKeyUtils.h"
+#include "kvstore/raftex/RaftPart.h"
+
+DEFINE_int32(snapshot_worker_threads, 4, "Threads number for snapshot");
+DEFINE_int32(snapshot_io_threads, 4, "Threads number for snapshot");
+DEFINE_int32(snapshot_send_retry_times, 3, "Retry times if send failed");
+DEFINE_int32(snapshot_send_timeout_ms, 60000, "Rpc timeout for sending snapshot");
+
+namespace nebula {
+namespace raftex {
+
+SnapshotManager::SnapshotManager() {
+    executor_.reset(new folly::IOThreadPoolExecutor(FLAGS_snapshot_worker_threads));
+    ioThreadPool_.reset(new folly::IOThreadPoolExecutor(FLAGS_snapshot_io_threads));
+}
+
+folly::Future<Status> SnapshotManager::sendSnapshot(std::shared_ptr<RaftPart> part,
+                                                    const HostAddr& dst) {
+    folly::Promise<Status> p;
+    auto fut = p.getFuture();
+    executor_->add([this, p = std::move(p), part, dst] () mutable {
+        auto spaceId = part->spaceId_;
+        auto partId = part->partId_;
+        auto termId = part->term_;
+        auto commitLogIdAndTerm = part->lastCommittedLogId();
```

Comment:
I think `commitLogIdAndTerm` may differs from the log term and id in `appendLogs` when we need to send snapshot, which is `committedLogId_` and `logTermToSend_`. Do you think there may be potential data lose?

Reply:
I think it is ok, because the log id is continuous even is commitLogIdAndTerm is different with real term and id.

Reply:
@critical27 You are right,   maybe the committedLogId is less than the real one in the snapshot.   It will not loss the data,  but maybe some record will be committed twice.   Let me add a TODO comment here firstly.

Reply:
I have understand the problem